File: yoonju/Programmers/Level 4/징검다리.js
function solution(distance, rocks, n) {
  // dist : 최소 거리값이어야 하는 거리
  const removeRock = (dist) => {
    let prevRock = 0;
    let deleteRock = 0;
    rocks.forEach((rock) => {
      // 현재 바위와 이전 바위의 거리가 dist보다 적다면 => 바위 제거
      if (rock - prevRock < dist) {
        deleteRock += 1;
      } else {
        // 현재 바위와 이전 바위의 거리가 dist보다 크거나 같다면 => 이전 바위 교체
        prevRock = rock;
      }
      // 제거한 바위가 n보다 많다 => 더 확인할 필요 없음
      if (deleteRock > n) {
        return;
      }
    });
    // 제거한 바위가 n보다 많으면 false, 같거나 적으면 true
    return deleteRock > n ? false : true;
  };
  var answer = 0;

  rocks.push(distance);

  rocks.sort((a, b) => a - b);

  let [s, e] = [1, distance];

  while (s <= e) {
    const mid = Math.floor((s + e) / 2);
    if (removeRock(mid)) {
      answer = mid;
      s = mid + 1;
    } else {
      e = mid - 1;
    }
  }

  return answer;
}

File: jieun/Programmers/징검다리.py
def solution(distance, rocks, n):
    rocks.sort()  # 오름차순 정렬
    rocks.append(distance)

    answer = 0
    l, r = 1, distance

    while l <= r:
        m = (l + r) // 2  # 바위 제거해서 최소 거리 m 만들 수 있는지?
        left_rock = 0  # 왼쪽 바위 위치
        cnt = 0  # 제거된 바위의 개수
        for i in range(len(rocks)):
            right_rock = rocks[i]  # 오른쪽 바위 위치
            if right_rock - left_rock < m:  # 간격이 m보다 작으면 오른쪽 바위(right_rock) 제거
                cnt += 1
            else:
                left_rock = right_rock  # 왼쪽 바위 위치를 옮긴다
            if n < cnt:
                break

        if cnt <= n:  # 돌을 더 제거해서 answer을 키울 수 있는가
            answer = m
            l = m + 1
        else:
            r = m - 1

    return answer


print(solution(25, [2, 14, 11, 21, 17], 2))  # 4
